operator/v1/types.go

```diff
@@ -0,0 +1,186 @@
+	Reason string `json:"reason,omitempty"`
+	// A human readable message indicating details about the transition.
+	Message string `json:"message,omitempty"`
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+}
+
+// JobConditionType defines all kinds of types of JobStatus.
+type JobConditionType string
+
+const (
+	// JobCreated means the job has been accepted by the system,
+	// but one or more of the pods/services has not been started.
+	// This includes time before pods being scheduled and launched.
+	JobCreated JobConditionType = "Created"
+
+	// JobRunning means all sub-resources (e.g. services/pods) of this job
+	// have been successfully scheduled and launched.
+	// The training is running without error.
+	JobRunning JobConditionType = "Running"
+
+	// JobRestarting means one or more sub-resources (e.g. services/pods) of this job
+	// reached phase failed but maybe restarted according to it's restart policy
+	// which specified by user in v1.PodTemplateSpec.
+	// The training is freezing/pending.
+	JobRestarting JobConditionType = "Restarting"
+
+	// JobSucceeded means all sub-resources (e.g. services/pods) of this job
+	// reached phase have terminated in success.
+	// The training is complete without error.
+	JobSucceeded JobConditionType = "Succeeded"
+
+	// JobFailed means one or more sub-resources (e.g. services/pods) of this job
+	// reached phase failed with no restarting.
+	// The training has failed its execution.
+	JobFailed JobConditionType = "Failed"
+)
+
+// CleanPodPolicy describes how to deal with pods when the job is finished.
+type CleanPodPolicy string
+
+const (
+	CleanPodPolicyUndefined CleanPodPolicy = ""
+	CleanPodPolicyAll       CleanPodPolicy = "All"
+	CleanPodPolicyRunning   CleanPodPolicy = "Running"
+	CleanPodPolicyNone      CleanPodPolicy = "None"
+)
+
+// RestartPolicy describes how the replicas should be restarted.
+// Only one of the following restart policies may be specified.
+// If none of the following policies is specified, the default one
+// is RestartPolicyAlways.
+type RestartPolicy string
+
+const (
+	RestartPolicyAlways    RestartPolicy = "Always"
+	RestartPolicyOnFailure RestartPolicy = "OnFailure"
+	RestartPolicyNever     RestartPolicy = "Never"
+
+	// `ExitCode` policy means that user should add exit code by themselves,
+	// The job operator will check these exit codes to
+	// determine the behavior when an error occurs:
+	// - 1-127: permanent error, do not restart.
+	// - 128-255: retryable error, will restart the pod.
+	RestartPolicyExitCode RestartPolicy = "ExitCode"
+)
+
+// RunPolicy encapsulates various runtime policies of the distributed training
+// job, for example how to clean up resources and how long the job can stay
+// active.
+type RunPolicy struct {
```

Comment:
I'm up for the idea of a new RunPolicy.  One question:  Once TFJob spec start to use this RunPolicy, it'll break the API compatibility. Is this the expectation ?

Reply:
The plan is to introduce another API version on TFJob. The current version (v1beta2) will not use this repo. 

Also we added support for multiple CRD versions in v0.5, We can also investigate how to use conversion webhooks for API compatibility: https://github.com/kubernetes/enhancements/issues/598. But looks like this feature requires k8s 1.15.